code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/SwiftRenderer.hs

```diff
@@ -942,52 +942,80 @@ swiftIndexFunc l v' = do
 
 swiftSplitFunc :: (RenderSym r) => Char -> SValue r -> SValue r
 swiftSplitFunc d s = let sepArg = var swiftSepBy char
   in objMethodCallNamedArgs (listType string) s swiftSplit [(sepArg, litChar d)]
 
 swiftJoinedFunc :: (RenderSym r) => Char -> SValue r -> SValue r
-swiftJoinedFunc d s = let sepArg = var swiftSep char 
+swiftJoinedFunc d s = let sepArg = var swiftSep char
   in objMethodCallNamedArgs string s swiftJoined [(sepArg, litChar d)]
 
 swiftIndexOf :: (RenderSym r) => SValue r -> SValue r -> SValue r
 swiftIndexOf = swiftUnwrapVal .: swiftIndexFunc
 
-swiftListSlice :: (RenderSym r) => SVariable r -> SValue r -> SValue r -> 
-  SValue r -> SValue r -> MSBlock r
-swiftListSlice vn vo beg end step = let i = var "i" int 
-  in block [vn &= swiftMapFunc (swiftStrideFunc beg end step) (lambda [i] 
-    (listAccess vo (valueOf i)))]
-
-swiftPrint :: Bool -> Maybe (SValue SwiftCode) -> SValue SwiftCode -> 
+-- | Swift's syntactic sugar for list slicing.
+--   All but the last two lines are used to manually set the bounds based on
+--   whether `step` is positive or negative, if either bound is not given
+swiftListSlice :: (RenderSym r) => SVariable r -> SValue r ->
+  Maybe (SValue r) -> Maybe (SValue r) -> SValue r -> MSBlock r
+swiftListSlice vn vo beg' end' step = do
+  stepV <- zoom lensMStoVS step
+  let needIf = (isNothing beg' || isNothing end') && isNothing (valueInt stepV)
+  bName <- if isJust beg' || not needIf then pure ""
+           else genVarName [] "begIdx"
+  eName <- if isJust end' || not needIf then pure ""
+           else genVarName [] "endIdx"
+  let -- If beg or end aren't given and step isn't a litInt, calculate them at
+      -- runtime and store in a variable.
+      begVar = var bName int
+      endVar = var eName int
+      findBounds = [setBeg | isNothing beg' && needIf] ++
+                 [setEnd | isNothing end' && needIf]
+      setBeg = varDecDef begVar (inlineIf (step ?> litInt 0) 
+                                (litInt 0) (listSize vo #- litInt 1))
+      setEnd = varDecDef endVar (inlineIf (step ?> litInt 0) 
+                                (listSize vo) (litInt (-1)))
+      -- If step is a litInt, use its value to compute beg and end directly
+      computedBeg = valueInt stepV >>= (\s -> if s > 0 then pure $ litInt 0
+                                              else pure $ listSize vo #- litInt 1)
+      computedEnd = valueInt stepV >>= (\s -> if s > 0 then pure $ listSize vo
+                                                       else pure $ litInt (-1))
+      -- Choose the values to use for beg and end based on given information
+      begVal = fromMaybe (fromMaybe (valueOf begVar) computedBeg) beg'
+      endVal = fromMaybe (fromMaybe (valueOf endVar) computedEnd) end'
+      i = var "i" int
+  block $ findBounds ++ [vn &= swiftMapFunc
+    (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))]
```

Comment:
After we spoke about this in person, I jumped down the rabbit hole... it can definitely still be cleaned up (better variable names and comments would go a long way in making it more readable), but I think I've de-duplicated all that I can now.

```haskell
swiftListSlice vn vo beg' end' step = do
  stepV <- zoom lensMStoVS step

  let (setBeg, begVal) = makeSetterVal "begIdx" (valueInt stepV) beg' (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" (valueInt stepV) end' (listSize vo) (litInt (-1))
      i = var "i" int

      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal vName Nothing _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
    makeSetterVal vName (Just s) Nothing  lb rb = 
      let theVar = var vName int
          computed = pure $ if s > 0 then lb else rb
      in (emptyStmt, fromMaybe (valueOf theVar) computed)
```

Here were my steps:
1. Bunch up the begX and endX code by beg and end.
2. Abstract those bunches of code into a local function.
3. Realized a small redundancy in how `needIf` is used, so I inlined it and simplified all expressions.

What do you think?

Reply:
Forgot to inline the last `makeSetterVal` case-set...

```haskell
swiftListSlice vn vo beg' end' step = do
  stepV <- zoom lensMStoVS step

  let (setBeg, begVal) = makeSetterVal "begIdx" (valueInt stepV) beg' (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" (valueInt stepV) end' (listSize vo) (litInt (-1))
      i = var "i" int

      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal vName Nothing _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
    makeSetterVal _    (Just s) Nothing  lb rb = (emptyStmt, if s > 0 then lb else rb)
```

Reply:
Ok, final time:

```haskell
swiftListSlice vn vo beg end step = do
  stepV <- zoom lensMStoVS step

  let mbStepV = valueInt stepV
      (setBeg, begVal) = makeSetterVal "begIdx" mbStepV beg (litInt 0)    (listSize vo #- litInt 1)
      (setEnd, endVal) = makeSetterVal "endIdx" mbStepV end (listSize vo) (litInt (-1))
      
      i = var "i" int
      setToSlice = vn &= swiftMapFunc (swiftStrideFunc begVal endVal step) (lambda [i] (listAccess vo (valueOf i)))
  block [
      setBeg,
      setEnd,
      setToSlice
    ]
  where
    makeSetterVal _     _       (Just v) _  _  = (emptyStmt, v)
    makeSetterVal _    (Just s) _        lb rb = (emptyStmt, if s > 0 then lb else rb)
    makeSetterVal vName _       _        lb rb = 
      let theVar = var vName int
          theSetter = varDecDef theVar $ inlineIf (step ?> litInt 0) lb rb
      in (theSetter, valueOf theVar)
```

Reply:
That looks great to me!  It'll also be handy soon when we make the fix for the `listSlice` in [Macros.hs](https://github.com/JacquesCarette/Drasil/blob/ab8043b24dbd6e5e4664013cf4a52d77d227956c/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/Macros.hs#L61-L78), that Java, C#, and C++ use.  We can move `makeSetterVal` to [CommonPseudoOO.hs](https://github.com/JacquesCarette/Drasil/blob/main/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/CommonPseudoOO.hs), and then both functions can use it!

Reply:
I tested it, and it passed all the test cases!  Would you like to push the commit, or should I?

Reply:
(spoke in person)